### ISSUE TITLE
Another last question about boot.jsm

### ISSUE THREAD

**Issue by Lepefe** [MrOtherGuy/fx-autoconfig#13]
Hey, please my last question about your newest boot.jsm.

I have this [js script](https://pastecode.io/s/ff9wmnjj). At the beginning of the script (heading) a function description is included, but in short, this js script is a replacement of the tab tooltip, it shows (inside urlbar) the url address of any mouse hovered tab. This script always worked like a charm, unfortunately until your newest boot.jsm. I have lot of js scripts, and this is the only script that didn't work with you newest boot.jsm.

Months ago I copied and pasted this js script from another user, days ago I tried to contact him, I hoped he may solve the problem (because he also uses your fx-autoconfig). But this user seems to be out of the grid.
So, this is not your responsible, but please I'll appreciate your help.

Interestingly, with your newest boot.jsm, this js script breaks the firefox window, and also interferes with other js scripts and css mods, it makes lot of chaos (when your newest boot.jsm is used). I opened this issue in order to request your help, but also, considering the chaos and the damage of the incompatibility, perhaps the info may be useful for your boot.jsm.

Thank you

**Comment by MrOtherGuy** [comment]
This is what I mentioned in [issue 12](https://github.com/MrOtherGuy/fx-autoconfig/issues/12#issuecomment-1112164560) about `gBrowser` which you are using quite a bit in that script.

I think the cleanest way would be to delay your function execution until startup has been finished, at which point gBrowser should be available.

You could do that for example like this:

1. Add `_ucUtils.startupFinished().then` to line 4
2. Remove the `()` at the end of the file. (the semi-colon can stay or not, it doesn't matter)

**Comment by Lepefe** [comment]
Thank you.
Your fix solved the problem. I'm going to test further and deeper to see if collateral damages appear. But at first sight, your fix works.

Yesterday when you explained the gBrowser issue, I searched all my js scripts, only two of them had the gBrowser.
One js script with the gBrowser is the one I shared here and you fixed. But before your gBrowser explanation, without knowing the gBrowser problem, I detected the incompatibility between my js script and your boot.jsm, and by myself I removed this js script. And even with this js script removed, for 3 days I couldn't make work your boot.jsm with my other js scripts. So, I know nothing, but it seems the gBrowser in this removed js script has no correlation with the problem I described in my https://github.com/MrOtherGuy/fx-autoconfig/issues/12
And as I said, I have no idea why yesterday, suddenly and mysteriously, all my other js scripts started to work with your newest boot.jsm.

I still have one js script using gBrowser.
It seems to work, I couldn't detect problems, and also while using your boot.jsm I couldn't detect interference between this js script and my other js scripts and css.
From my point of view, everything seems to works.

But it doesn't hurt to ask and to learn: Should I change the gBrowser in this js script?
For those js scripts using gBrowser, is there a kind of general fix that can be applied?
The fix in your message above can be used for all scripts having gBrowser?

I hope my questions may help future users having incompatibilities with gBrowser.

**Comment by MrOtherGuy** [comment]
Actually, what I said above is not completely accurate. It seems that in your script the thing causing all the mayhem is that you are trying to access `gURLBar`. Now, I'm not exactly certain about this, but I believe that internally gURLBar expects that gBrowser *is* made available before it, but in this case it is not.

You can test that by doing the following edits to your script:

1. Add this after line 5: `const gBrowser = window.gBrowser || window._gBrowser;`
2. Replace `gURLBar` on line 75 with `document.getElementById("urlbar")`

This achieves two things: variable named `gBrowser` now exists in your function scope and is reference to either window.gBrowser or window._gBrowser - whichever is available. Also, gURLBar is now never accessed, and thus it won't break internally. Instead you just get reference to urlbar which shouldn't break everything even if window.gBrowser is not around.

I'm debating if I should just revert acb9460 to add the gBrowser hack back in. I would rather not, but I'm not sure what the best course of action would be here.

**Comment by Lepefe** [comment]
Yeap, your second patch works.
Just to be clear, I'm not using your two patches. I took my original script I shared with you (without your first patch), I applied your second patch, and my original js script works (using only your second patch).
That's great. Thank you.

With regards to gBrowser, as I told you, I personally don't have any visible problems, so for me everything is working.
But just for the sake of learning, if you don't revert gBrowser hack, do we users need to change something in our scripts? Is there an universal fix? (Is the copy and paste of https://github.com/MrOtherGuy/fx-autoconfig/commit/acb9460ce608dae4e5531dac04f3921ba990da04 a solution inside our scripts using gBrowser? Where to paste?).
And if I have a script with gBrowser, but if it is not showing incompatibilities, do I need to meed with the gBrowser in this script?

**Comment by Lepefe** [comment]
Or another option could be to keep your newest boot.jsm (without gBrowser hack), because is the right one, and you're right, the best boot.jsm should always be kept in your repo. At the same time, if it's not too much trouble for you, in parallel you may offer a second boot.jsm with the gBrowser hack, but temporarily, until you decide which boot.jsm is the best for you repo, or until devs adapt their scripts to your fx-autoconfig.
At any case, in the name of your fx-autoconfig users, I thank you for your great work, I'm sure lot of silent users are using it, and we are tremendously enjoying the benefits of your work. Thank you!

**Comment by MrOtherGuy** [comment]
But I should note that the reason why your script broke everything is because it tried to use access gURLBar very early. Apparently it does work, but your script is causing urlbar to be constructed "much" earlier in the window startup than what Firefox would expect. It shouldn't be surprising that such thing can break things.

For that reason I think the ideal solution for your script would be to wait until the window is ready, which was what I suggested at first. I mean, you would never need to show your custom urlbar "title" before window has been created anyway so you might as well wait few milliseconds.

**Comment by MrOtherGuy** [comment]
I could add a pref that users could use to enable/disable the hack though...

**Comment by MrOtherGuy** [comment]
Sidenote: This is not really about the same issue anymore. But since you're here and using nightlies I felt to inform you that probably in the near-ish future there will be a need to update both program/config.js as well as boot.jsm to change how scripts - including boot.jsm itself - are imported. I'm not sure how much stuff is going to break but there's definitely going to be need to *some* changes.

**Comment by Lepefe** [comment]
The pref userChromeJS.gBrowser_hack.enabled works perfect (at least for me). Zero problems, exactly as you described, if I use my script without your patch, then everything is broken, then I turn ON the new pref, then restart, and everything works (no need to clean startup cache folder), everything simple, solved in seconds. A beautiful solution. Congratulations and thank you.

> Sidenote: This is not really about the same issue anymore. But since you're here and using nightlies I felt to inform you that probably in the near-ish future there will be a need to update both program/config.js as well as boot.jsm to change how scripts - including boot.jsm itself - are imported.

Thank you for alerting me.
IMHO changes always are welcome, because most of the time changes mean improvements. Yeah, it's a kind of headache to fix stuff after updates or upgrades. But everyone dealing with customizations must accept this part of the game.
And people like you, always doing a great work, keeping their scripts updated and upgraded, facilitates the assimilation or adoption of new changes. You're anonymous heroes. Thank you Man!
I only fear radical changes like blocking css or autoconfig customization, but I hope this is not the case.

**Comment by MrOtherGuy** [comment]
Alright, in [4a047ac](https://github.com/MrOtherGuy/fx-autoconfig/commit/4a047acb099ee01a02c3d440eb99d4fd43eaaba5) I added mechanism to show the user a message if boot.jsm detects broken startup.

It would be very cool if you could test and confirm that the following indeed does happen:

Prerequisite: set the earlier pref `userChromeJS.gBrowser_hack.enabled` to `false` and clear startup cache.

1. Launch Firefox - it should break on startup as before.
2. After about 5s, a message should appear  saying things are broken
3. Click "Enable workaround" in the message
4. Firefox restarts and should now work
5. Message should appear
6. Click "Why am I seeing this" -button in the message
7. A new tab should open in this repository's readme error section
8. If you go to about:config, both `userChromeJS.gBrowser_hack.enabled` and `userChromeJS.gBrowser_hack.required` should be `true`

Sidenote, simply trying to access `gBrowser` in a custom script should not "break" anything - it should just produce error in browser console. This new mechanism doesn't detect that and thus no additional message is shown. It's only if a serious startup breakage is detected that the message is shown.

**Comment by Lepefe** [comment]
It works like a charm.
Beautiful experience.
Thank you.

It took me lot of time and testing. I was experiencing a firefox startup delay with your new boot.jsm, and I thought it was the culprit. But it wasn't.
Yesterday my antivirus received an upgrade, and I dunno why, but now it's creating a small delay in my firewall startup. It's detecting nothing, just creates a small delay. But it took me lot of tests and time to discover that.
There is no relation between the AV upgrade and your fx-autoconfig, it seems to be something related just to the new AV upgrade itself.

The bottom line is that your new boot,jsm works perfectly.
I tested your boot.jsm using my script (with lot of gBrowser elements), and the behavior was 100% as you described. In another clean test, I tested my script with your first patch, and everything worked like a charm. I tested your second patch, zero problems, in all tests your newest boot.jsm worked perfectly.
Using gBrowser or patches or scripts without gBrowsers, your boot.jsm worked for me. I couldn't find a single problem (and my Browser Console is clean).

Congrats.

**Comment by MrOtherGuy** [comment]
That is very cool. Great, thanks! I'll mark this issue solved then
